packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -21,6 +22,361 @@ import 'gesture_detector.dart';
+  /// and we would prefer to use the identical [DiagnosticsNode] for each time
+  /// a node exists in the path.
+  final List<DiagnosticsNode> children;
+
+  /// Index of the child that the path continues on.
+  ///
+  /// Equal to `null` if the path does not continue.
+  final int childIndex;
+
+  /// Creates a full description of a step in a path through a tree of
+  /// [DiagnosticsNode] objects.
+  ///
+  /// The [child] argument must not be null.
+  DiagnosticsPathNode({ this.node, this.children, this.childIndex }) : assert(children != null);
+}
+
+List<DiagnosticsPathNode> _followDiagnosticableChain(List<Diagnosticable> chain, {
+  String name,
+  DiagnosticsTreeStyle style,
+}) {
+  final List<DiagnosticsPathNode> path = <DiagnosticsPathNode>[];
+  if (chain.isEmpty)
+    return path;
+  DiagnosticsNode diagnostic = chain.first.toDiagnosticsNode(name: name, style: style);
+  for (int i = 1; i < chain.length; i += 1) {
+    final Diagnosticable target = chain[i];
+    bool foundMatch = false;
+    final List<DiagnosticsNode> children = diagnostic.getChildren();
+    for (int j = 0; j < children.length; j += 1) {
+      final DiagnosticsNode child = children[j];
+      if (child.value == target) {
+        foundMatch = true;
+        path.add(new DiagnosticsPathNode(
+          node: diagnostic,
+          children: children,
+          childIndex: j,
+        ));
+        diagnostic = child;
+        break;
+      }
+    }
+    assert(foundMatch);
+  }
+  path.add(new DiagnosticsPathNode(node: diagnostic, children: diagnostic.getChildren()));
+  return path;
+}
+
+typedef void InspectorSelectionChangedCallback();
+
+/// Structure to help reference count Dart objects referenced by a GUI tool
+/// using [WidgetInspectorService].
+class _InspectorReferenceData {
+  _InspectorReferenceData(this.object);
+
+  final Object object;
+  int count = 1;
+}
+
+/// Service used by GUI tools to interact with the [WidgetInspector].
+///
+/// Calls to this object are typically made from GUI tools such as the [Flutter
+/// IntelliJ Plugin](https://github.com/flutter/flutter-intellij/blob/master/README.md)
+/// using the [Dart VM Service protocol](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md).
+/// This class uses its own object id and manages object lifecycles itself
+/// instead of depending on the [object ids](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#getobject)
+/// specified by the VM Service Protocol because the VM Service Protocol ids
+/// expire unpredictably. Object references are tracked in groups so that tools
+/// that clients can use dereference all objects in a group with a single
+/// operation making it easier to avoid memory leaks.
+///
+/// All methods returning String values return JSON.
+class WidgetInspectorService {
+  /// The current [WidgetInspectorService].
+  static final WidgetInspectorService instance = new WidgetInspectorService();
```

Comment:
you don't have a private constructor, so anyone can create a new instance of this; is that intentional?

Reply:
Creating unique instances is handy for testing. Is there a pattern you would prefer for Flutter members that are "public for testing"?

Reply:
If it's possible to have two instances and have them work sanely together, then we should just remove the instance member and have the consumers of this class create their own instance, IMHO.

Reply:
This singleton should be the source of truth of which widgets are inspected. The IntelliJ Plugin is relying on WidgetInspectorService.instance existing but I can change that if you feel strongly.
https://github.com/flutter/flutter-intellij/blob/master/src/io/flutter/inspector/InspectorService.java#L117

Reply:
I think we should either have a singleton API or an API that can handle multiple simultaneous instances. I don't really mind which. I think we should not have an API that can only handle one instance but doesn't prevent you from having multiple instances at once.

One option is to have a `dispose()` method that neuters the instance (such that it can't be used again) and resets the static. The constructor would then throw if the instance variable was not null, and would set the instance variable to itself if it was. That might work for the tests, while still ensuring there's only one instance for "real" work.

Reply:
Made the constructor private and just made the api a singleton. Added a disposeAllGroups method to make the tests robust as that is sufficient and less heavy weight than adding a dispose method that neuters the instance.